extensions/amp-web-push/0.1/amp-web-push-config.js

```diff
@@ -0,0 +1,204 @@
+ * have the ID "amp-web-push". Subscribe and unsubscribe actions dispatched from
+ * various <amp-web-push-widgets> are all processed by <amp-web-push-config>
+ * which then forwards the event to the web push service.
+ */
+export class WebPushConfig extends AMP.BaseElement {
+
+  /** @param {!AmpElement} element */
+  constructor(element) {
+    super(element);
+  }
+
+  /** @override */
+  isLayoutSupported(layout) {
+    return layout == Layout.NODISPLAY;
+  }
+
+  validate() {
+    this.ensureSpecificElementId_();
+    this.ensureUniqueElement_();
+
+    const config = {
+      helperIframeUrl: null,
+      permissionDialogUrl: null,
+      serviceWorkerUrl: null,
+    };
+
+    for (const attribute in WebPushConfigAttributes) {
+      const value = WebPushConfigAttributes[attribute];
+      user().assert(this.element.getAttribute(value),
+          `The ${value} attribute is required for <${CONFIG_TAG}>`);
+      config[value] = this.element.getAttribute(value);
+    }
+
+    if (!this.isValidHelperOrPermissionDialogUrl_(config['helperIframeUrl'])) {
+      throw user().createError(`<${CONFIG_TAG}> must have a valid ` +
+        'helperIframeUrl attribute. It should begin with ' +
+        'the https:// protocol and point to the provided lightweight ' +
+        'template page provided for AMP messaging.');
+    }
+
+    if (!this.isValidHelperOrPermissionDialogUrl_(
+        config['permissionDialogUrl'])) {
+      throw user().createError(`<${CONFIG_TAG}> must have a valid ` +
+        'permissionDialogUrl attribute. It should begin with ' +
+        'the https:// protocol and point to the provided template page ' +
+        'for showing the permission prompt.');
+    }
+
+    if (parseUrl(config['serviceWorkerUrl']).protocol !== 'https:') {
+      throw user().createError(`<${CONFIG_TAG}> must have a valid ` +
+        'serviceWorkerUrl attribute. It should begin with the ' +
+        'https:// protocol and point to the service worker JavaScript file ' +
+        'to be installed.');
+    }
+
+    if (parseUrl(config['serviceWorkerUrl']).origin !==
+          parseUrl(config['permissionDialogUrl']).origin ||
+        parseUrl(config['permissionDialogUrl']).origin !==
+        parseUrl(config['helperIframeUrl']).origin) {
+      throw user().createError(`<${CONFIG_TAG}> URL attributes ` +
+        'serviceWorkerUrl, permissionDialogUrl, and ' +
+        'helperIframeUrl must all share the same origin.');
+    }
+  }
+
+  /**
+   * Ensures this element is defined with TAG id.
+   *
+   * @private
+   */
+  ensureSpecificElementId_() {
+    if (this.element.getAttribute('id') !== TAG) {
+      throw user().createError(`<${CONFIG_TAG}> must have an id ` +
+        'attribute with value \'' + TAG + '\'.');
+    }
+  }
+
+  /**
+   * Ensures there isn't another page element with the same id.
+   * @private
+   */
+  ensureUniqueElement_() {
+    if (this.getAmpDoc().win.document.querySelectorAll('#' + TAG).length > 1) {
```

Comment:
It should be called at the right time, e.g. in or after `buildCallback` and it'll work then. You have to use root node in ampdoc, b/c otherwise querying `win.document.body` will not work. E.g. in case when an ampdoc is in a shadow root, `win.document.querySelector()` won't be able to penetrate shadow root to look there. Let's find a way to get `getAmpDoc()` working here.